Sources/ClientRuntime/Networking/Streaming/BufferedStream.swift

```diff
@@ -0,0 +1,128 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+
+/// A `Stream` implementation that buffers data in memory.
+/// The buffer size depends on the amount of data written and read.
+/// Note: This class is thread-safe.
+/// Note: if data is not read from the stream, the buffer will grow indefinitely until the stream is closed.
+///       or reach the maximum size of a `Data` object.
+public class BufferedStream: Stream {
+
+    /// Returns the length of the stream, if known
+    public private(set) var length: Int?
```

Comment:
We kind of touched on this but I forget where we ended up. 
Does this need to be optional? Doesn't every stream start with a length of 0 and gets incremented as more data is written to the stream. Can a stream never have a length? What does an unknown length mean?

Reply:
Not every stream has length. This becomes more prominent when event streams comes in picture. We don't know how many events customer is suppose to send.

Based on its nullability, we decide the content length header

https://github.com/awslabs/smithy-swift/blob/jangirg/refactor/streams/Sources/ClientRuntime/Networking/Http/Middlewares/ContentLengthMiddleware.swift#L23

Reply:
Hmm. I don't have full context but length feels like the wrong mechanism to dictate the headers. It would be much simpler to make length non-null, and use a value of 0 in the nil cases. This makes things much easier to reason about. 

Perhaps we should add a property to stream to dictate the encoding headers (`streamType` or `preferredHeader`), or the middleware maps the stream to the proper header type (if event stream -> use x header else use y header). That would also make things much easier reason about as it isn't clear, from reading the code alone, why length would dictate the header value.

Reply:
per our offline chat, length is nullable, however we stopped setting it to 0 now.